nixos/modemmanager: support additional FCC unlock scripts

flokli opened the pull request:
https://github.com/NixOS/nixpkgs/pull/155414 introduced an option to support enabling the FCC unlock scripts that ModemManager provides, but doesn't execute (since 1.18.4). However, it wasn't possible to add additional, provided-by-the-vendor unlock scripts.

This commit fixes this up, by renaming the option from `networking.networkmanager.enableFccUnlock` to
`networking.networkmanager.enableBundledFccUnlockScripts`, and adding an additional option `networking.networkmanager.fccUnlockScripts`, which allows configuring (other) unlock scripts.

The logic is used in case
`networking.networkmanager.enableBundledFccUnlockScripts` is set to true - it'll then pick up the list of exposed
modemmanager.passthru.fccUnlockScripts and set/extend `networking.networkmanager.fccUnlockScripts` with them.

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc @jwygoda

flokli committed:
nixos/modemmanager: support additional FCC unlock scripts

This commit introduces a `networking.networkmanager.fccUnlockScripts`
option, which allows specifying additional, usually vendor-provided
unlock scripts.

networking.networkmanager.enableBundledFccUnlockScripts is refactored to
make use of the same mechanism internally.

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2205.section.md:
```diff
@@ -936,9 +936,10 @@ In addition to numerous new and upgraded packages, this release has the followin
-  [services.networking.networkmanager.enableFccUnlock](#opt-networking.networkmanager.enableFccUnlock)
-  was added to support FCC unlock procedures. Since release 1.18.4, the ModemManager
-  daemon no longer automatically performs the FCC unlock procedure by default. See
+  [services.networking.networkmanager.enableBundledFccUnlockScripts](#opt-networking.networkmanager.enableBundledFccUnlockScripts)
+  was added to support FCC unlock procedures.
+  Since release 1.18.4, the ModemManager daemon no longer automatically performs
+  the FCC unlock procedure by default. See
```
We shouldn't modify old release notes.

jtojnar replied:
This is required in order not to break link checker.

flokli replied:
I removed the hyperlink, put it in backticks, and squashed the two lines together.

SuperSandro2000 commented on nixos/modules/services/networking/networkmanager.nix:
```diff
@@ -510,6 +537,16 @@ in {
 
+      # if cfg.enableBundledFccUnlockScripts is set, populate
+      # networking.networkmanager.fccUnlockScripts with the values from
+      # pkgs.modemmanager.passthru.fccUnlockScripts.
```
We don't have networking.networkmanager.package?

flokli replied:
The unlock scripts are provided by modemmanager, not NetworkManager, even though it's in the "networkmanager" config namespace.

Moving this into a separate namespace is out of scope for this PR.

SuperSandro2000 commented on nixos/modules/services/networking/networkmanager.nix:
```diff
@@ -369,16 +369,38 @@ in {
+          Enable FCC unlock procedures shipped with ModemManager.
+          Since release 1.18.4, the ModemManager daemon no longer
           automatically performs the FCC unlock procedure by default. See
           [the docs](https://modemmanager.org/docs/modemmanager/fcc-unlock/)
```
```suggestion
          [the modemmanager docs](https://modemmanager.org/docs/modemmanager/fcc-unlock/)
```

flokli replied:
This is unrelated to this PR and was already present before.

jtojnar commented on pkgs/tools/networking/modemmanager/default.nix:
```diff
@@ -93,6 +94,21 @@ stdenv.mkDerivation rec {
+      "2c7c:030a" = "${modemmanager}/share/ModemManager/fcc-unlock.available.d/2c7c";
+      "413c:81a3" = "${modemmanager}/share/ModemManager/fcc-unlock.available.d/1199";
+      "413c:81a8" = "${modemmanager}/share/ModemManager/fcc-unlock.available.d/1199";
+    };
```
Please add a check to `passthru.tests` ensuring the code is kept in sync (see e. g. NetworkManager or fwupd packages).

flokli replied:
I don't think modemmanager will include more unlock scripts ever. It looks like that vendors will provide their own custom unlock scripts, rather than playing this whack-a-mole game, as least when reading https://modemmanager.org/docs/modemmanager/fcc-unlock/, which also lists these as an (exhaustive) list.

I'd personally prefer dropping all of `enableBundledFccUnlockScripts`, and leave it up to the user to explicitly set their unlock script (in case more SAR-aware scripts pop up somewhere) over trying to make this update-proof. I only kept this for backwards-compat with the previous NixOS option.

RaitoBezarius replied:
I agree with @flokli on this one, @jtojnar do you think this is a blocking matter?

flokli replied:
Just to make sure, my /preference/ would be to entirely drop the auto enabling of the bundled FCC unlock scripts, and document how to enable the bundled ones in the release notes. I'm happy to update the PR to that, but we need to decide which direction to go.

jtojnar replied:
Sorry, I thought I have already replied that I am fine with removing the `enableBundledFccUnlockScripts` but apparently not.

flokli replied:
No problem, thanks for confirming. I'll update the PR.

jtojnar commented on nixos/modules/services/networking/networkmanager.nix:
```diff
@@ -387,7 +409,8 @@ in {
     (mkRenamedOptionModule [ "networking" "networkmanager" "useDnsmasq" ] [ "networking" "networkmanager" "dns" ])
-    (mkRemovedOptionModule ["networking" "networkmanager" "dynamicHosts"] ''
+    (mkRenamedOptionModule [ "networking" "networkmanager" "enableFccUnlock" ] [ "networking" "networkmanager" "enableBundledFccUnlockScripts" ])
+    (mkRemovedOptionModule [ "networking" "networkmanager" "dynamicHosts" ] ''
```
Could you please extract the cleanups to a separate commit?

flokli replied:
I can split the rename into a commit before the actual change, sure.

flokli replied:
pushed.

flokli committed:
nixos/modemmanager: enableFccUnlock -> enableBundledFccUnlockScripts

PR #155414 introduced an option to support enabling the FCC unlock
scripts that ModemManager provides, but since 1.18.4 doesn't execute
anymore.

However, this option is specifically only about the unlock scripts
provided with ModemManager so far. Rename the option to make this more
obvious.

flokli committed:
nixos/networkmanager: nixpkgs-fmt

flokli commented:
@jtojnar poke ;-)

As written previously, I think having the user explicitly choose which unlock script to use (so the `networking.networkmanager.fccUnlockScripts = [{id = "vid:pid"; path = "…"}]` path) over the "enable all unlock scripts known route.

We could also drop the `enableBundledFccUnlockScripts` option alltogether, and explain how to add modemmanager-provided unlock scripts to the release notes. 

This is more in line with the upstream documentation, where users expliticly do symlinks for a pid:vid of their choosing, if they decide so.

RaitoBezarius commented:
Result of `nixpkgs-review pr 252283` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>

flokli committed:
nixos/modemmanager: remove enableBundledFccUnlockScripts option

This removes the networking.networkmanager.enableBundledFccUnlockScripts
option, and updates the release notes.

flokli commented:
As discussed in https://github.com/NixOS/nixpkgs/pull/252283#discussion_r1323418885, I updated the PR to now remove the networking.networkmanager.enableFccUnlock script entirely. PTAL.
